sep/026: Add task to fix sepolia fp upgrade

sebastianst opened the pull request:
**Description**

Adds task to fix the Sepolia FP upgrade, since the old one (https://github.com/ethereum-optimism/superchain-ops/pull/374) referenced a wrongly newly deployed `DelayedWETH` implementation.

~Also adds a framework for creating reusable verification scripts.~ The framework was split into its own PR with #429 

**Additional context**

~This task may actually be obsolete because of #410~ update: it's necessary because #410 reused the wrong `DelayedWETH`s.

Based on https://github.com/ethereum-optimism/superchain-ops/pull/442 which adds additional checks to DG upgrades.

**Metadata**

Towards https://github.com/ethereum-optimism/optimism/issues/13033

sebastianst commented on justfile:
```diff
@@ -4,7 +4,7 @@ install: install-contracts install-eip712sign
 install-contracts:
   #!/usr/bin/env bash
-  forge install
+  forge install --shallow
```
might speed up `forge install` significantly

mds1 replied:
I would say we leave this alone for now and keep the plain `forge install`. I can't imagine this change being a problem, but we have a lot of moving parts with all the ongoing playbooks and we should minimize the diff/potential sources of error as much as possible, just to be safe

sebastianst commented on .circleci/config.yml:
```diff
@@ -20,6 +20,28 @@ parameters:
     type: integer
     default: 5
 
+commands:
+  simulate_nested:
+    description: "Runs simulations of a nested task"
+    parameters:
+      task:
+        type: string
+    steps:
+      - checkout
+      - run:
+          name: "simulate nested << parameters.task >>"
+          command: |
+            just install
+            cd tasks/<< parameters.task >>
+            SIMULATE_WITHOUT_LEDGER=true just \
+              --dotenv-path $(pwd)/.env \
+              --justfile ../../../nested.just \
+              simulate foundation
+            SIMULATE_WITHOUT_LEDGER=true just \
+              --dotenv-path $(pwd)/.env \
+              --justfile ../../../nested.just \
+              simulate council
```
we should leave this one in for future tasks, which can then just run this command instead of copy-pasting the simulation invocation, see below `simulate_sep_025`

mds1 replied:
nice, I like this

Inphi commented:
fyi the 025 task number is used and now executed by https://github.com/ethereum-optimism/superchain-ops/pull/410

mds1 commented on tasks/sep/025-fp-holocene-upgrade-fix/NestedSignFromJson.s.sol:
```diff
@@ -0,0 +1,34 @@
+import {CouncilFoundationNestedSign } from "script/verification/CouncilFoundationNestedSign.s.sol";
+import {SuperchainRegistry} from "script/verification/Verification.s.sol";
+
+contract NestedSignFromJson is OriginalNestedSignFromJson, CouncilFoundationNestedSign, DisputeGameUpgrade {
+    constructor() SuperchainRegistry("sepolia", "op", "v1.8.0-rc.4")
+    DisputeGameUpgrade(
```
Thank you for writing these reusable validations, this is really nice

sebastianst replied:
Glad you're ok with the design!

mds1 commented on tasks/sep/025-fp-holocene-upgrade-fix/README.md:
I have stopped reviewing here, since in slack you mentioned the README and VALIDATION still needs to be adapted. But directionally I really like this!

sebastianst replied:
It's ready now.

mds1 commented on NESTED-VALIDATION.md:
````diff
@@ -0,0 +1,102 @@
+The GnosisSafe `approvedHashes` mapping is updated to indicate approval of this transaction by the chosen Safe Role. The correctness of this slot can be verified as follows:
+- Since this is a nested mapping, we need to use `cast index` twice to confirm that this is the correct slot. The inputs needed are:
+    - The location (`8`) of the `approvedHashes` mapping in the [GnosisSafe storage layout](https://github.com/safe-global/safe-contracts/blob/v1.4.0/contracts/libraries/SafeStorage.sol#L23)
+    - The address of the Safe Role, stored at the env var `$SAFE_ROLE` in the following cast script command.
+    - The safe hash to approve, stored at the env var `$SAFE_HASH` in the following cast script command.
+      It's the value after "Nested hash:" in the simulation output logs.
+- Then using `cast index`, we can compute the key with
+    ```shell
+      $ cast index bytes32 $SAFE_HASH $(cast index address $SAFE_ROLE 8)
+    ```
+    The output of this command must match the key of the state change.
````
Is this suggestion accurate? Safe Role was a bit unclear to me
```suggestion
    - The address of the Safe signer, stored at the env var `$SAFE_SIGNER` in the following cast script command.
    - The safe hash to approve, stored at the env var `$SAFE_HASH` in the following cast script command.
      It's the value after "Nested hash:" in the simulation output logs.
- Then using `cast index`, we can compute the key with
    ```shell
      $ cast index bytes32 $SAFE_HASH $(cast index address $SAFE_SIGNER 8)
    ```
    The output of this command must match the key of the state change.
```

mds1 replied:
This is not a blocker, feel free to ignore this change, because I see in https://github.com/ethereum-optimism/superchain-ops/pull/425 we use the same terminology and I don't want to create unnecessary rework

sebastianst replied:
I've introduced a definition of `Safe Role` further up in the doc. But I'm happy to change the definition to `Safe Signer`.

mds1 commented on tasks/sep/026-fp-holocene-upgrade-fix/VALIDATION.md:
````diff
@@ -0,0 +1,49 @@
+- Council simulation: `0x1188ef4fb2355c49d442a7f950dad4492a2961a34786ce407c3db8c8d331250d`
+- Foundation simulation: `0x030c89f6f0b759d892697ad05de88183a3f6f579e0abe30c81427eaefb20a459`
+
+calculated as explained in the nested validation doc:
+```sh
+SAFE_HASH=0x572c29de03270811431295a31c25273c7dbf3bc0f17f29a8fe3131f86b7a9cb6 # "Nested hash:"
+SAFE_ROLE=0xf64bc17485f0B4Ea5F06A96514182FC4cB561977 # Council
+cast index bytes32 $SAFE_HASH $(cast index address $SAFE_ROLE 8)
+# 0x1188ef4fb2355c49d442a7f950dad4492a2961a34786ce407c3db8c8d331250d
+
+SAFE_ROLE=0xDEe57160aAfCF04c34C887B5962D0a69676d3C8B # Foundation
+cast index bytes32 $SAFE_HASH $(cast index address $SAFE_ROLE 8)
+# 0x030c89f6f0b759d892697ad05de88183a3f6f579e0abe30c81427eaefb20a459
+```
````
same as above
```suggestion
calculated as explained in the nested validation doc:
```sh
SAFE_HASH=0x572c29de03270811431295a31c25273c7dbf3bc0f17f29a8fe3131f86b7a9cb6 # "Nested hash:"
SAFE_SIGNER=0xf64bc17485f0B4Ea5F06A96514182FC4cB561977 # Council
cast index bytes32 $SAFE_HASH $(cast index address $SAFE_SIGNER 8)
# 0x1188ef4fb2355c49d442a7f950dad4492a2961a34786ce407c3db8c8d331250d

SAFE_SIGNER=0xDEe57160aAfCF04c34C887B5962D0a69676d3C8B # Foundation
cast index bytes32 $SAFE_HASH $(cast index address $SAFE_SIGNER 8)
# 0x030c89f6f0b759d892697ad05de88183a3f6f579e0abe30c81427eaefb20a459
```
```

sebastianst replied:
renamed to Safe Signer

Ethnical commented on tasks/sep/026-fp-holocene-upgrade-fix/VALIDATION.md:
```diff
@@ -0,0 +1,49 @@
+- **Key**: `0xffdfc1249c027f9191656349feb0761381bb32c9f557e01f419fd08754bf5a1b` <br/>
+  **Before**: `0x0000000000000000000000005e0877a8f6692ed470013e651c4357d0c4941e6c` <br/>
+  **After**: `0x000000000000000000000000e591ebbc2ba0ead3db6a0867cc132fe1c123f448` <br/>
+  **Meaning**: Updates the CANNON game type implementation. You can verify which implementation is set using `cast call 0x05F9613aDB30026FFd634f38e5C4dFd30a197Fa1 "gameImpls(uint32)(address)" 0`, where `0` is the [`CANNON` game type](https://github.com/ethereum-optimism/optimism/blob/op-contracts/v1.4.0/packages/contracts-bedrock/src/dispute/lib/Types.sol#L28).
```
I think to be consistent we should have the same convention naming. 

```suggestion
  **Meaning**: Updates the PERMISSIONELESS_CANNON game type implementation. You can verify which implementation is set using `cast call 0x05F9613aDB30026FFd634f38e5C4dFd30a197Fa1 "gameImpls(uint32)(address)" 0`, where `0` is the [`CANNON` game type](https://github.com/ethereum-optimism/optimism/blob/op-contracts/v1.4.0/packages/contracts-bedrock/src/dispute/lib/Types.sol#L28).
```

sebastianst replied:
Ah I see the confusion. We call the _permissionless_ version just `CANNON`, that's how [the enum is called in the contracts](https://github.com/ethereum-optimism/optimism/blob/op-contracts/v1.4.0/packages/contracts-bedrock/src/dispute/lib/Types.sol#L28).

Ethnical replied:
Thanks make sense to me!

Ethnical commented on tasks/sep/026-fp-holocene-upgrade-fix/OVERVIEW.md:
```diff
@@ -0,0 +1,37 @@
+**_impl:** `0xb51baD2d9Da9f94d6A4A5A493Ae6469005611B68`
+
+
+## Tx #2: Upgrade `CANNON` game type in `DisputeGameFactory`
```
```suggestion
## Tx #2: Upgrade `PERMISSIONLESS_CANNON` game type in `DisputeGameFactory`
```

Ethnical commented on tasks/sep/026-fp-holocene-upgrade-fix/README.md:
```diff
@@ -0,0 +1,37 @@
+
+## Pre-deployments
+
+- `MIPS` - `0x69470D6970Cd2A006b84B1d4d70179c892cFCE01`
```
In this upgrade, is the MIPS contract is used? 
I can't see any transaction or storage modified with it?

sebastianst commented:
I just realized that this task may have become obsolete by https://github.com/ethereum-optimism/superchain-ops/pull/410 - in this case, I'll change this task to only add the verification framework, but will delete the task sep/026 itself. @mds1 @Inphi @mbaxter

Ethnical commented on tasks/sep/026-fp-holocene-upgrade-fix/input.json:
```diff
@@ -0,0 +1,67 @@
+  "chainId": 11155111,
+  "metadata": {
+    "name": "Holocene Hardfork - Proof Contract Upgrades",
+    "description": "Upgrades the `MIPS.sol`, `FaultDisputeGame.sol`, and `PermissionedDisputeGame.sol` contracts for Holocene."
```
where is the MIPS upgrade? I can't see any storage referring to this contract?

sebastianst replied:
MIPS is implicitly upgraded by upgrading the games. The games reference the new MIPS implementation. This is also part of the validations. There's no MIPS proxy contract that needs upgrading, it's a singleton that is referenced directly from DG contracts.

sebastianst commented:
Rebased on top of the framework-only PR https://github.com/ethereum-optimism/superchain-ops/pull/429 and squashed commits. Reduces the diff to only the actual task 026.

sebastianst commented:
Going to switch the FDG and PDG implementations to new MT-variants that @mbaxter deployed
* FDG: `0x833a817ef459f4ecdb83fc5a4bf04d09a4e83f3f`
* PDG: `0xbbd576128f71186a0f9ae2f2aab4afb4af2dae17`

Ethnical commented on tasks/sep/026-fp-holocene-upgrade-fix/.env:
```diff
@@ -0,0 +1,12 @@
+FOUNDATION_SAFE=0xDEe57160aAfCF04c34C887B5962D0a69676d3C8B
+OWNER_SAFE=0x1Eb2fFc903729a0F03966B917003800b145F56E2
+SAFE_NONCE=""
+SIMULATE_WITHOUT_LEDGER=0 # 1
```
This is causing issue with the justfile since the justfile is checking only if the `SIMULATE_WITHOUT_LEDGER` is set and not the value of the var.

Ethnical commented on tasks/sep/026-fp-holocene-upgrade-fix/NestedSignFromJson.s.sol:
```diff
@@ -0,0 +1,57 @@
+    constructor()
+        SuperchainRegistry("sepolia", "op", "v1.8.0-rc.4")
+        DisputeGameUpgrade(
+            0x03b7eaa4e3cbce90381921a4b48008f4769871d64f93d113fcadca08ecee503b, // absolutePrestate
```
I have verified against the MT FDG.  https://sepolia.etherscan.io/address/0x833a817eF459f4eCdB83Fc5A4Bf04d09A4e83f3F#readContract
But I don't know if there is another place to get the information from?

Inphi replied:
It's the mt64 prestate for the `op-program/v1.4.0-rc.3` release. Which supports op-sepolia holocene.
Documented here - https://www.notion.so/oplabs/OP-Sepolia-v1-9-0-rc-3-MT-Cannon-15af153ee16280e58414dbed970d43ef.

Ethnical replied:
Thaanks!

Ethnical commented on tasks/sep/026-fp-holocene-upgrade-fix/VALIDATION.md:
```diff
@@ -0,0 +1,49 @@
+### `0x05F9613aDB30026FFd634f38e5C4dFd30a197Fa1` (`DisputeGameFactoryProxy`)
+
+- **Key**: `0xffdfc1249c027f9191656349feb0761381bb32c9f557e01f419fd08754bf5a1b` <br/>
+  **Before**: `0x0000000000000000000000005e0877a8f6692ed470013e651c4357d0c4941e6c` <br/>
```
The value in the before in the before is incorrect:
```suggestion
  **Before**: `0x000000000000000000000000924d3d3b3b16e74bab577e50d23b2a38990dd52c` <br/>
```

Ethnical commented on tasks/sep/026-fp-holocene-upgrade-fix/VALIDATION.md:
```diff
@@ -0,0 +1,49 @@
+  Before this task has been executed, you will see that the returned address is `0x0000000000000000000000005e0877a8f6692ed470013e651c4357d0c4941e6c`, matching the "Before" value of this slot, demonstrating this slot is storing the address of the CANNON implementation.
+
+- **Key**: `0x4d5a9bd2e41301728d41c8e705190becb4e74abe869f75bdb405b63716a35f9e` <br/>
+  **Before**: `0x0000000000000000000000004ed046e66c96600dae1a4ec39267bb0ce476e8cc` <br/>
```
The value in the `before` is incorrect: 
```suggestion
  **Before**: `0x000000000000000000000000879e899523ba9a4ab212a2d70cf1af73b906cbe5` <br/>
```

Ethnical commented on tasks/sep/026-fp-holocene-upgrade-fix/VALIDATION.md:
```diff
@@ -0,0 +1,49 @@
+  **Before**: `0x0000000000000000000000005e0877a8f6692ed470013e651c4357d0c4941e6c` <br/>
+  **After**: `0x000000000000000000000000833a817ef459f4ecdb83fc5a4bf04d09a4e83f3f` <br/>
+  **Meaning**: Updates the CANNON game type implementation. You can verify which implementation is set using `cast call 0x05F9613aDB30026FFd634f38e5C4dFd30a197Fa1 "gameImpls(uint32)(address)" 0`, where `0` is the [`CANNON` game type](https://github.com/ethereum-optimism/optimism/blob/op-contracts/v1.4.0/packages/contracts-bedrock/src/dispute/lib/Types.sol#L28).
+  Before this task has been executed, you will see that the returned address is `0x0000000000000000000000005e0877a8f6692ed470013e651c4357d0c4941e6c`, matching the "Before" value of this slot, demonstrating this slot is storing the address of the CANNON implementation.
```
Here, also we need to update the value accordingly. 
![CleanShot 2025-01-07 at 23 47 18@2x](https://github.com/user-attachments/assets/97a9f8e9-a763-40c7-b848-72b7a4b39387)

Ethnical commented on tasks/sep/026-fp-holocene-upgrade-fix/VALIDATION.md:
```diff
@@ -0,0 +1,49 @@
+  **Before**: `0x0000000000000000000000004ed046e66c96600dae1a4ec39267bb0ce476e8cc` <br/>
+  **After**: `0x000000000000000000000000bbd576128f71186a0f9ae2f2aab4afb4af2dae17` <br/>
+  **Meaning**: Updates the PERMISSIONED_CANNON game type implementation. You can verify which implementation is set using `cast call 0x05F9613aDB30026FFd634f38e5C4dFd30a197Fa1 "gameImpls(uint32)(address)" 1`, where `1` is the [`PERMISSIONED_CANNON` game type](https://github.com/ethereum-optimism/optimism/blob/op-contracts/v1.4.0/packages/contracts-bedrock/src/dispute/lib/Types.sol#L31).
+  Before this task has been executed, you will see that the returned address is `0x0000000000000000000000004ed046e66c96600dae1a4ec39267bb0ce476e8cc`, matching the "Before" value of this slot, demonstrating this slot is storing the address of the PERMISSIONED_CANNON implementation.
```
Here, also we need to update the value accordingly with the correct game address in the before (matching the screen).
<img width="1824" alt="image" src="https://github.com/user-attachments/assets/5580b1b3-a53e-4909-a46a-e9ad4b5ceb60" />

Ethnical reviewed (commented):
Nice work! We need to fix the validation file before once this is done we can merge directly!
Sorry for the delay!
